### ISSUE TITLE
[Enhancement] Nfo files

### ISSUE THREAD

**Issue by OmgImAlexis** [beetbox/beets#2413]
It'd be great to see an option for NFO files being written for artists and albums which would mean the MBID can be saved as well as other details. This helps with things like KODI, Plex, etc. and would also help with beets and updating music libraries even if the database file is deleted.

For an example of the NFO files have a look at what KODI accepts. Most other platforms that use NFO files for music either use the same format as KODI or an extended format which KODI's as a base. 

http://kodi.wiki/view/NFO_files/Music
http://kodi.wiki/view/Music_tagging (go down to "6 NFO files")

**Comment by akyag** [comment]
+1000000000000
And there is no other software which can do this for music. And maybe in the future artist artwork, thumbs, backgrounds from https://fanart.tv/music-fanart/  for ex. can be fetched too?

**Comment by jrobeson** [comment]
@akyag: there's already a fanart.tv backend for the fetchart plugin, so that part is done

**Comment by jrobeson** [comment]
@akyag : sorry, i thought we did have fanart.tv already, but this implies not so https://github.com/beetbox/beets/issues/889

**Comment by sampsyo** [comment]
Ah, so these are little XML files containing metadata? Seems like a fine request.

**Comment by tigranl** [comment]
Hi, @sampsyo! Where can we store .nfo files? I think I can implement this feature.

**Comment by sampsyo** [comment]
Cool! I actually don't know where Kodi expects them to be... perhaps in the album directory (`album.path` in beets)?

**Comment by tigranl** [comment]
Another one question. I assume it should be a plugin?

**Comment by sampsyo** [comment]
Yeah, totally—integrations with specific other software count as narrow enough use cases that they should be plugins.

**Comment by bajosoto** [comment]
Hi @tigranl. I'm also interested in implementing this feature. If you have already started something in a repo would you mind sharing so maybe we can collaborate on this? Also, do you think this kind of feature would be more useful as a CLI command, or as part of the importing process?

**Comment by bajosoto** [comment]
And to your question, it's two kinds of file. artist.nfo goes in the artist folder and album.nfo in the album one. Also, according to the wiki, "Both album.nfo and artist.nfo files can contain the MusicBrainz tag or simply contain a URL to the correct page on the MusicBrainz website". Maybe this would be a much simpler approach than generating the xml structure?

**Comment by tigranl** [comment]
@bajosoto I just started, I'll make a PR soon. I think it would be more appropriate as part of the importing process.

**Comment by tigranl** [comment]
Yeah, using just a URL would simplify everything.

**Comment by bajosoto** [comment]
Ok, thanks. I have a very, very simple and rough (but working!) implementation of a plugin [here](https://github.com/bajosoto/kodinfo-beets) that generates the .nfo file for albums (no artists yet) as part of the import process like you mentioned. I am aware I'm doing a second call to music brainz that could (and should) be spared since this is already part of the import process, but I'm not quite sure of how can I pull the data that is already in the task in commands.py (the url is already fetched there) all the way into my plugin. It's probably a silly question, but I'm a beginner with python and any input or feedback would be much appreciated. :D

**Comment by sampsyo** [comment]
Cool, thanks for getting this started!

It's actually pretty simple to get the ID out of the album object. You can just use `album.mb_albumid` to get the MBID that has been assigned to the album after matching.

If you're curious about the full set of fields that are available on `album`, it's [here](https://github.com/beetbox/beets/blob/17ad3e83dbe20126ff0048c29e1c1dd245845645/beets/library.py#L871-L903).

**Comment by bajosoto** [comment]
How did I miss that..? Thanks @sampsyo. I'll improve it tomorrow :bowtie:

**Comment by tigranl** [comment]
It might be a good idea to use 'import' event, so it will work with singletones as well.

**Comment by akyag** [comment]
how would it work for singletons? one nfo for every single track?

**Comment by tigranl** [comment]
Yep.

**Comment by tigranl** [comment]
https://gist.github.com/tigranl/79652e7dfac1d6216c1c0a2d0bbf696f

**Comment by bajosoto** [comment]
Hey @tigranl, from what I see in the [KODI wiki](http://kodi.wiki/view/NFO_files/Music), there seems to be no template for singletons so these would probably not be recognized by it (perhaps this works in Plex or others?). Nevertheless they can be generated. I believe using the item_imported event would be a good idea, alongside album_import since it's not called during album imports from what I understand from the documentation.
@sampsyo, is there any way to fetch the album artist path from within the plugin based on the path configuration?

**Comment by sampsyo** [comment]
There's no specific album associated with each album artist. However, you might try using the parent directory of the album, which is likely to work with *most* (but not all) path configurations.

**Comment by bajosoto** [comment]
Yeah, that's what I was thinking. I guess it wouldn't work for certain configurations.. Thanks!

**Comment by peace899** [comment]
I have been thinking about creating .nfo files for kodi using beets data but unfortunately I'm not clued up on databases and python. But I made headway by creating a bash script. Found [here](https://github.com/peace899/nfokodi)

**Comment by peace899** [comment]
With my limited python knowledge and lots of googling, I came up with this [script](https://github.com/peace899/beets2kodi/blob/master/beets2kodi.py). Obviously would like to turn it into a plugin (or run it with [runafter.py plugin](https://github.com/m-urban/beets/blob/5fc5b0836e354085860277d9f3bc71dfa9d4141b/beetsplug/runafter.py), any help welcome!

**Comment by peace899** [comment]
**RE: artist path**
I have made this attempt to get artist path from the config file. Probably might work for most configurations with $albumartist in default path:

```
import os, re
import beets.library
from beets import config

libpath = os.path.expanduser(str(config['library']))
lib = beets.library.Library(libpath)
mb_albumid = 'e26afcc2-7ef9-49e9-b64f-8f74a5ff15c3' #example mbid
albumid = 'mb_albumid:'+ mb_albumid

def artist_path(albumid):
    for album in lib.albums(albumid):
        albumpath = album.path.decode("utf-8")
        albumartist = album.albumartist
        folder = str(config['directory']) + '/' + str(config['paths']['default'])
        config_items = list(folder.split('/'))
        indices = [i for i, s in enumerate(config_items) if 'albumartist' in s]
        x = int(indices[0])
        albumpath_items = list(albumpath.split('/'))[1:x]
        artist_path = "/" + '/'.join(map(str, albumpath_items)) + '/' + albumartist
        return artist_path
 
```
